main.go

```diff
@@ -216,3 +227,46 @@ func registerCollectors(registry prometheus.Registerer, kubeClient clientset.Int
 
 	glog.Infof("Active collectors: %s", strings.Join(activeCollectors, ","))
 }
+
+type GathererFunc func() ([]*dto.MetricFamily, error)
+
+func (f GathererFunc) Gather() ([]*dto.MetricFamily, error) {
+	return f()
+}
+
+func filteredGatherer(r prometheus.Gatherer, whitelist options.MetricSet, blacklist options.MetricSet) prometheus.Gatherer {
+	whitelistEnabled := !whitelist.IsEmpty()
+	blacklistEnabled := !blacklist.IsEmpty()
+	filterEnabled := whitelistEnabled || blacklistEnabled
+
+	return GathererFunc(func() ([]*dto.MetricFamily, error) {
```

Comment:
I suggest we make this function a package-level function thus we can add some unit tests for it. We need to determine the relationship between whitelist and blacklist:
* Does we permit users specify both whitelist and blacklist
* Does kube-state-metrics should only prefer whitelist and ignore blacklist even blacklist is not empty. 

@brancz 

Currently, the implementation is kube-state-metrics prefer whitelist metrics even blacklist is not empty which i think is not the right logic. :)